gradle-plugin/src/main/java/com/google/cloud/tools/dependencies/gradle/LinkageCheckTask.java

```diff
@@ -191,59 +197,162 @@ private String dependencyPathsOfProblematicJars(
-      String dependencyPath =
-          stack.stream().map(this::formatComponentResult).collect(Collectors.joining(" / "));
-      output.put(coordinates, dependencyPath);
+  /** Dependency nodes to record dependency paths while traversing the dependency tree */
+  private static class ResolvedComponentResultNode {
+    ResolvedComponentResult componentResult;
+
+    ResolvedComponentResultNode parent;
+
+    ResolvedComponentResultNode(
+        ResolvedComponentResult componentResult, ResolvedComponentResultNode parent) {
+      this.componentResult = componentResult;
+      this.parent = parent;
+    }
+
+    boolean hasParent(ResolvedComponentResult other) {
+      if (componentResult.equals(other)) {
+        return true;
+      }
+      if (parent == null) {
+        return false;
+      }
+      return parent.hasParent(other);
+    }
+
+    public String pathFromRoot() {
+      List<String> dependencyPathElementsReversed = new ArrayList<>();
+      for (ResolvedComponentResultNode iter = this; iter != null; iter = iter.parent) {
+        dependencyPathElementsReversed.add(formatComponentResult(iter.componentResult));
+      }
+      Collections.reverse(dependencyPathElementsReversed);
+      String dependencyPath = Joiner.on(" / ").join(dependencyPathElementsReversed);
+      return dependencyPath;
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(this)
+          .add("componentResult", componentResult)
+          .add("parent", parent)
+          .toString();
     }
 
-    for (DependencyResult dependencyResult : item.getDependencies()) {
-      if (dependencyResult instanceof ResolvedDependencyResult) {
-        ResolvedDependencyResult resolvedDependencyResult =
-            (ResolvedDependencyResult) dependencyResult;
-        ResolvedComponentResult child = resolvedDependencyResult.getSelected();
+    @Override
+    public boolean equals(Object other) {
+      if (this == other) {
+        return true;
+      }
+      if (other == null || getClass() != other.getClass()) {
+        return false;
+      }
+      ResolvedComponentResultNode that = (ResolvedComponentResultNode) other;
+      return Objects.equals(componentResult, that.componentResult)
+          && Objects.equals(parent, that.parent);
+    }
 
-        if (stack.contains(child)) {
-          // Circular dependency check
-          if (checkedCircularDependency.add(child)) {
+    @Override
+    public int hashCode() {
+      return Objects.hash(componentResult, parent);
+    }
+  }
+
+  /**
+   * Returns mapping from Maven coordinates to their dependency paths appearing in the dependency
+   * graph
+   *
+   * @param componentResult The root project in the dependency graph
+   * @param targetCoordinatesSet The Maven coordinates to check their dependency paths
+   */
+  private ListMultimap<String, String> groupCoordinatesToDependencyPaths(
+      ResolvedComponentResult componentResult, Set<String> targetCoordinatesSet) {
```

Comment:
Is `componentResult` expected to be the root of dependency tree? Then maybe we could name it this way. Or I could be confused...

Reply:
Updated to rootProject.